File: apps/api/internal/service/auth_service.go
package service

import (
	"context"
	"fmt"
	"regexp"
	"strings"
	apperrors "survey-api/internal/errors"
	"survey-api/internal/model"
	"survey-api/internal/repository"
	"time"

	"github.com/golang-jwt/jwt/v5"
	"golang.org/x/crypto/bcrypt"
)

type AuthService interface {
	Register(ctx context.Context, req *RegisterRequest) (*AuthResponse, error)
	Login(ctx context.Context, req *LoginRequest) (*AuthResponse, error)
	RefreshToken(ctx context.Context, refreshToken string) (*AuthResponse, error)
	Logout(ctx context.Context, tokenID string) error
	GetProfile(ctx context.Context, userID uint64) (*UserProfile, error)
	UpdateProfile(ctx context.Context, userID uint64, req *UpdateProfileRequest) error
}

type authService struct {
	userRepo  repository.UserRepository
	jwtSecret string
}

type RegisterRequest struct {
	Email    string `json:"email" validate:"required,email"`
	Password string `json:"password" validate:"required,min=6"`
	Name     string `json:"name" validate:"required,min=2"`
}

type LoginRequest struct {
	Email    string `json:"email" validate:"required,email"`
	Password string `json:"password" validate:"required"`
}

type UpdateProfileRequest struct {
	Name string `json:"name" validate:"required,min=2"`
}

type AuthResponse struct {
	User         *UserProfile `json:"user"`
	AccessToken  string       `json:"access_token"`
	RefreshToken string       `json:"refresh_token"`
	ExpiresIn    int64        `json:"expires_in"`
}

type UserProfile struct {
	ID        uint64    `json:"id"`
	UUID      string    `json:"uuid"`
	Email     string    `json:"email"`
	Name      string    `json:"name"`
	IsActive  bool      `json:"is_active"`
	CreatedAt time.Time `json:"created_at"`
}

func NewAuthService(userRepo repository.UserRepository, jwtSecret string) AuthService {
	return &authService{
		userRepo:  userRepo,
		jwtSecret: jwtSecret,
	}
}

func (s *authService) Register(ctx context.Context, req *RegisterRequest) (*AuthResponse, error) {
	// Validate input
	if err := s.validateRegisterRequest(req); err != nil {
		return nil, err
	}

	// Check if user already exists
	existingUser, err := s.userRepo.GetByEmail(ctx, req.Email)
	if err == nil && existingUser != nil {
		apperrors.LogError(fmt.Errorf("user already exists"), map[string]interface{}{
			"email":     req.Email,
			"operation": "register",
		})
		return nil, apperrors.ErrUserExists
	}

	// Hash password
	hashedPassword, err := bcrypt.GenerateFromPassword([]byte(req.Password), bcrypt.DefaultCost)
	if err != nil {
		apperrors.LogError(err, map[string]interface{}{
			"operation": "password_hash",
		})
		return nil, apperrors.Wrap(err, apperrors.ErrCodeInternal, "Failed to hash password", 500)
	}

	// Create user
	user := &model.User{
		Email:    strings.ToLower(strings.TrimSpace(req.Email)),
		Password: string(hashedPassword),
		Name:     strings.TrimSpace(req.Name),
		IsActive: true,
	}

	if err := s.userRepo.Create(ctx, user); err != nil {
		apperrors.LogError(err, map[string]interface{}{
			"email":     req.Email,
			"operation": "user_create",
		})
		return nil, apperrors.DatabaseError("user creation", err)
	}

	// Generate tokens
	accessToken, refreshToken, expiresIn, err := s.generateTokens(user.ID, user.UUID)
	if err != nil {
		apperrors.LogError(err, map[string]interface{}{
			"user_id":   user.ID,
			"operation": "token_generation",
		})
		return nil, apperrors.Wrap(err, apperrors.ErrCodeInternal, "Failed to generate tokens", 500)
	}

	apperrors.LogInfo("User registered successfully", map[string]interface{}{
		"user_id": user.ID,
		"email":   user.Email,
	})

	return &AuthResponse{
		User:         s.toUserProfile(user),
		AccessToken:  accessToken,
		RefreshToken: refreshToken,
		ExpiresIn:    expiresIn,
	}, nil
}

func (s *authService) Login(ctx context.Context, req *LoginRequest) (*AuthResponse, error) {
	// Validate input
	if err := s.validateLoginRequest(req); err != nil {
		return nil, err
	}

	// Get user by email
	user, err := s.userRepo.GetByEmail(ctx, strings.ToLower(strings.TrimSpace(req.Email)))
	if err != nil {
		apperrors.LogError(err, map[string]interface{}{
			"email":     req.Email,
			"operation": "login",
		})
		return nil, apperrors.ErrInvalidCredentials
	}

	// Check if user is active
	if !user.IsActive {
		apperrors.LogWarn("Inactive user attempted login", map[string]interface{}{
			"user_id": user.ID,
			"email":   user.Email,
		})
		return nil, apperrors.ErrUserInactive
	}

	// Verify password
	if err := bcrypt.CompareHashAndPassword([]byte(user.Password), []byte(req.Password)); err != nil {
		apperrors.LogWarn("Invalid password attempt", map[string]interface{}{
			"user_id": user.ID,
			"email":   user.Email,
		})
		return nil, apperrors.ErrInvalidCredentials
	}

	// Generate tokens
	accessToken, refreshToken, expiresIn, err := s.generateTokens(user.ID, user.UUID)
	if err != nil {
		apperrors.LogError(err, map[string]interface{}{
			"user_id":   user.ID,
			"operation": "token_generation",
		})
		return nil, apperrors.Wrap(err, apperrors.ErrCodeInternal, "Failed to generate tokens", 500)
	}

	apperrors.LogInfo("User logged in successfully", map[string]interface{}{
		"user_id": user.ID,
		"email":   user.Email,
	})

	return &AuthResponse{
		User:         s.toUserProfile(user),
		AccessToken:  accessToken,
		RefreshToken: refreshToken,
		ExpiresIn:    expiresIn,
	}, nil
}

func (s *authService) RefreshToken(ctx context.Context, refreshToken string) (*AuthResponse, error) {
	// Parse refresh token
	token, err := jwt.Parse(refreshToken, func(token *jwt.Token) (interface{}, error) {
		if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
			return nil, fmt.Errorf("unexpected signing method: %v", token.Header["alg"])
		}
		return []byte(s.jwtSecret), nil
	})

	if err != nil || !token.Valid {
		return nil, apperrors.ErrTokenInvalid
	}

	claims, ok := token.Claims.(jwt.MapClaims)
	if !ok {
		return nil, apperrors.ErrTokenInvalid
	}

	userUUID, ok := claims["user_uuid"].(string)
	if !ok {
		return nil, apperrors.ErrTokenInvalid
	}

	// Get user
	user, err := s.userRepo.GetByUUID(ctx, userUUID)
	if err != nil {
		return nil, apperrors.ErrUserNotFound
	}

	// Generate new tokens
	accessToken, newRefreshToken, expiresIn, err := s.generateTokens(user.ID, user.UUID)
	if err != nil {
		return nil, fmt.Errorf("failed to generate tokens: %w", err)
	}

	return &AuthResponse{
		User:         s.toUserProfile(user),
		AccessToken:  accessToken,
		RefreshToken: newRefreshToken,
		ExpiresIn:    expiresIn,
	}, nil
}

func (s *authService) Logout(ctx context.Context, tokenID string) error {
	// In a real implementation, you would invalidate the token in database
	// For now, we'll just return success
	return nil
}

func (s *authService) GetProfile(ctx context.Context, userID uint64) (*UserProfile, error) {
	user, err := s.userRepo.GetByID(ctx, userID)
	if err != nil {
		apperrors.LogError(err, map[string]interface{}{
			"user_id":   userID,
			"operation": "get_profile",
		})
		return nil, apperrors.ErrUserNotFound
	}

	return s.toUserProfile(user), nil
}

func (s *authService) UpdateProfile(ctx context.Context, userID uint64, req *UpdateProfileRequest) error {
	// Validate input
	if err := s.validateUpdateProfileRequest(req); err != nil {
		return err
	}

	user, err := s.userRepo.GetByID(ctx, userID)
	if err != nil {
		apperrors.LogError(err, map[string]interface{}{
			"user_id":   userID,
			"operation": "update_profile",
		})
		return apperrors.ErrUserNotFound
	}

	user.Name = strings.TrimSpace(req.Name)
	if err := s.userRepo.Update(ctx, user); err != nil {
		apperrors.LogError(err, map[string]interface{}{
			"user_id":   userID,
			"operation": "update_profile",
		})
		return apperrors.DatabaseError("user update", err)
	}

	apperrors.LogInfo("User profile updated", map[string]interface{}{
		"user_id": userID,
		"name":    user.Name,
	})

	return nil
}

func (s *authService) generateTokens(userID uint64, userUUID string) (string, string, int64, error) {
	// Access token (15 minutes)
	accessClaims := jwt.MapClaims{
		"user_id":   userID,
		"user_uuid": userUUID,
		"exp":       time.Now().Add(15 * time.Minute).Unix(),
		"type":      "access",
	}

	accessToken := jwt.NewWithClaims(jwt.SigningMethodHS256, accessClaims)
	accessTokenString, err := accessToken.SignedString([]byte(s.jwtSecret))
	if err != nil {
		return "", "", 0, err
	}

	// Refresh token (7 days)
	refreshClaims := jwt.MapClaims{
		"user_uuid": userUUID,
		"exp":       time.Now().Add(7 * 24 * time.Hour).Unix(),
		"type":      "refresh",
	}

	refreshToken := jwt.NewWithClaims(jwt.SigningMethodHS256, refreshClaims)
	refreshTokenString, err := refreshToken.SignedString([]byte(s.jwtSecret))
	if err != nil {
		return "", "", 0, err
	}

	return accessTokenString, refreshTokenString, 15 * 60, nil
}

func (s *authService) toUserProfile(user *model.User) *UserProfile {
	return &UserProfile{
		ID:        user.ID,
		UUID:      user.UUID,
		Email:     user.Email,
		Name:      user.Name,
		IsActive:  user.IsActive,
		CreatedAt: user.CreatedAt,
	}
}

// Validation functions
func (s *authService) validateRegisterRequest(req *RegisterRequest) error {
	if req.Email == "" {
		return apperrors.ValidationError("email", "Email is required")
	}
	if req.Name == "" {
		return apperrors.ValidationError("name", "Name is required")
	}
	if req.Password == "" {
		return apperrors.ValidationError("password", "Password is required")
	}

	// Email validation
	if !isValidEmail(req.Email) {
		return apperrors.ValidationError("email", "Invalid email format")
	}

	// Name validation
	if len(strings.TrimSpace(req.Name)) < 2 {
		return apperrors.ValidationError("name", "Name must be at least 2 characters")
	}
	if len(strings.TrimSpace(req.Name)) > 100 {
		return apperrors.ValidationError("name", "Name must be less than 100 characters")
	}

	// Password validation
	if len(req.Password) < 6 {
		return apperrors.ValidationError("password", "Password must be at least 6 characters")
	}
	if len(req.Password) > 128 {
		return apperrors.ValidationError("password", "Password must be less than 128 characters")
	}

	return nil
}

func (s *authService) validateLoginRequest(req *LoginRequest) error {
	if req.Email == "" {
		return apperrors.ValidationError("email", "Email is required")
	}
	if req.Password == "" {
		return apperrors.ValidationError("password", "Password is required")
	}

	// Email validation
	if !isValidEmail(req.Email) {
		return apperrors.ValidationError("email", "Invalid email format")
	}

	return nil
}

func (s *authService) validateUpdateProfileRequest(req *UpdateProfileRequest) error {
	if req.Name == "" {
		return apperrors.ValidationError("name", "Name is required")
	}

	// Name validation
	if len(strings.TrimSpace(req.Name)) < 2 {
		return apperrors.ValidationError("name", "Name must be at least 2 characters")
	}
	if len(strings.TrimSpace(req.Name)) > 100 {
		return apperrors.ValidationError("name", "Name must be less than 100 characters")
	}

	return nil
}

// Helper functions
func isValidEmail(email string) bool {
	emailRegex := regexp.MustCompile(`^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$`)
	return emailRegex.MatchString(email)
}

File: src/lib/auth-service.ts
import { 
  IAuthService, 
  IUserRepository, 
  IPasswordService, 
  IJwtService,
  RegisterRequest, 
  LoginRequest, 
  RefreshTokenRequest,
  AuthResponse, 
  AuthUser,
  ValidationError,
  AuthenticationError,
  UserAlreadyExistsError,
  InvalidRefreshTokenError
} from '@/types/auth';

/**
 * Authentication service implementation
 * Follows SOLID principles - Single Responsibility Principle and Dependency Inversion Principle
 */
export class AuthService implements IAuthService {
  constructor(
    private userRepository: IUserRepository,
    private passwordService: IPasswordService,
    private jwtService: IJwtService
  ) {}

  /**
   * Register a new user
   * @param request - Registration request data
   * @returns Auth response with user and token
   */
  async register(request: RegisterRequest): Promise<AuthResponse> {
    try {
      // Validate input
      this.validateRegisterRequest(request);

      // Check if user already exists
      const existingUser = await this.userRepository.findByEmail(request.email);
      if (existingUser) {
        throw new UserAlreadyExistsError('User with this email already exists');
      }

      // Hash password
      const passwordHash = await this.passwordService.hashPassword(request.password);

      // Create user
      const user = await this.userRepository.create({
        email: request.email,
        passwordHash,
        name: request.name || null,
        isActive: true,
        refreshToken: null,
        refreshTokenExp: null,
      });

      // Generate token
      const token = this.jwtService.generateToken({
        userId: user.id,
        email: user.email,
      });

      // Return response
      return {
        success: true,
        data: {
          user: this.mapUserToAuthUser(user),
          token,
        },
      };
    } catch (error) {
      if (error instanceof ValidationError || 
          error instanceof UserAlreadyExistsError) {
        throw error;
      }
      
      throw new Error(`Registration failed: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  /**
   * Login user
   * @param request - Login request data
   * @returns Auth response with user and token
   */
  async login(request: LoginRequest): Promise<AuthResponse> {
    try {
      // Validate input
      this.validateLoginRequest(request);

      // Find user by email
      const user = await this.userRepository.findByEmail(request.email);
      if (!user) {
        throw new AuthenticationError('Invalid email or password');
      }

      // Check if user is active
      if (!user.isActive) {
        throw new AuthenticationError('Account is deactivated');
      }

      // Verify password
      const isPasswordValid = await this.passwordService.verifyPassword(
        request.password,
        user.passwordHash
      );

      if (!isPasswordValid) {
        throw new AuthenticationError('Invalid email or password');
      }

      // Generate tokens
      const token = this.jwtService.generateToken({
        userId: user.id,
        email: user.email,
      });

      const refreshToken = this.jwtService.generateRefreshToken({
        userId: user.id,
        email: user.email,
      });

      // Store refresh token in database
      const refreshTokenExp = new Date();
      refreshTokenExp.setDate(refreshTokenExp.getDate() + 7); // 7 days from now

      await this.userRepository.update(user.id, {
        refreshToken,
        refreshTokenExp,
      });

      // Return response
      return {
        success: true,
        data: {
          user: this.mapUserToAuthUser(user),
          token,
          refreshToken,
        },
      };
    } catch (error) {
      if (error instanceof ValidationError || 
          error instanceof AuthenticationError) {
        throw error;
      }
      
      throw new Error(`Login failed: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  /**
   * Refresh access token using refresh token
   * @param request - Refresh token request data
   * @returns Auth response with new tokens
   */
  async refreshToken(request: RefreshTokenRequest): Promise<AuthResponse> {
    try {
      // Validate input
      this.validateRefreshTokenRequest(request);

      // Verify refresh token
      const payload = this.jwtService.verifyRefreshToken(request.refreshToken);
      if (!payload) {
        throw new InvalidRefreshTokenError('Invalid or expired refresh token');
      }

      // Find user
      const user = await this.userRepository.findById(payload.userId);
      if (!user || !user.isActive) {
        throw new InvalidRefreshTokenError('Invalid or expired refresh token');
      }

      // Check if stored refresh token matches
      if (user.refreshToken !== request.refreshToken) {
        throw new InvalidRefreshTokenError('Invalid or expired refresh token');
      }

      // Check if refresh token is expired
      if (user.refreshTokenExp && user.refreshTokenExp < new Date()) {
        throw new InvalidRefreshTokenError('Invalid or expired refresh token');
      }

      // Generate new tokens
      const newToken = this.jwtService.generateToken({
        userId: user.id,
        email: user.email,
      });

      const newRefreshToken = this.jwtService.generateRefreshToken({
        userId: user.id,
        email: user.email,
      });

      // Update refresh token in database
      const refreshTokenExp = new Date();
      refreshTokenExp.setDate(refreshTokenExp.getDate() + 7); // 7 days from now

      await this.userRepository.update(user.id, {
        refreshToken: newRefreshToken,
        refreshTokenExp,
      });

      // Return response
      return {
        success: true,
        data: {
          user: this.mapUserToAuthUser(user),
          token: newToken,
          refreshToken: newRefreshToken,
        },
      };
    } catch (error) {
      if (error instanceof ValidationError || 
          error instanceof InvalidRefreshTokenError) {
        throw error;
      }
      
      throw new Error(`Token refresh failed: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  /**
   * Revoke refresh token for a user
   * @param userId - User ID
   */
  async revokeRefreshToken(userId: number): Promise<void> {
    try {
      await this.userRepository.update(userId, {
        refreshToken: null,
        refreshTokenExp: null,
      });
    } catch (error) {
      throw new Error(`Failed to revoke refresh token: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  /**
   * Logout user by revoking refresh token
   * @param userId - User ID
   * @returns Success response
   */
  async logout(userId: number): Promise<{ success: boolean; message: string }> {
    try {
      await this.revokeRefreshToken(userId);
      return {
        success: true,
        message: 'Logout successful',
      };
    } catch (error) {
      throw new Error(`Logout failed: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  /**
   * Validate token and return user
   * @param token - JWT token
   * @returns User data or null if invalid
   */
  async validateToken(token: string): Promise<AuthUser | null> {
    try {
      if (!token) {
        return null;
      }

      // Verify token
      const payload = this.jwtService.verifyToken(token);
      if (!payload) {
        return null;
      }

      // Find user
      const user = await this.userRepository.findById(payload.userId);
      if (!user || !user.isActive) {
        return null;
      }

      return this.mapUserToAuthUser(user);
    } catch (error) {
      console.error('Token validation failed:', error);
      return null;
    }
  }

  /**
   * Validate registration request
   * @param request - Registration request
   */
  private validateRegisterRequest(request: RegisterRequest): void {
    if (!request.email) {
      throw new ValidationError('Email is required');
    }

    if (!request.password) {
      throw new ValidationError('Password is required');
    }

    if (!this.isValidEmail(request.email)) {
      throw new ValidationError('Invalid email format');
    }

    // Basic password validation
    if (!request.password || request.password.length < 6) {
      throw new ValidationError('Password must be at least 6 characters long');
    }

    if (request.name && request.name.length > 100) {
      throw new ValidationError('Name must be less than 100 characters');
    }
  }

  /**
   * Validate login request
   * @param request - Login request
   */
  private validateLoginRequest(request: LoginRequest): void {
    if (!request.email) {
      throw new ValidationError('Email is required');
    }

    if (!request.password) {
      throw new ValidationError('Password is required');
    }

    if (!this.isValidEmail(request.email)) {
      throw new ValidationError('Invalid email format');
    }
  }

  /**
   * Validate refresh token request
   * @param request - Refresh token request
   */
  private validateRefreshTokenRequest(request: RefreshTokenRequest): void {
    if (!request.refreshToken) {
      throw new ValidationError('Refresh token is required');
    }

    if (typeof request.refreshToken !== 'string' || request.refreshToken.trim().length === 0) {
      throw new ValidationError('Invalid refresh token format');
    }
  }

  /**
   * Validate email format
   * @param email - Email to validate
   * @returns True if valid email format
   */
  private isValidEmail(email: string): boolean {
    const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    return emailRegex.test(email);
  }

  /**
   * Map Prisma User to AuthUser (exclude sensitive data)
   * @param user - Prisma User object
   * @returns AuthUser object
   */
  private mapUserToAuthUser(user: { id: number; email: string; name: string | null; isActive: boolean; createdAt: Date; updatedAt: Date }): AuthUser {
    return {
      id: user.id,
      email: user.email,
      name: user.name || undefined,
      isActive: user.isActive,
      createdAt: user.createdAt,
      updatedAt: user.updatedAt,
    };
  }
}
